airflow/hooks/fs_hooks/s3.py

```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+
+from builtins import super
+
+import posixpath
+
+import s3fs
+
+from .base import FsHook
+
+
+class S3FsHook(FsHook):
+    """Hook for interacting with files in S3."""
+
+    def __init__(self, conn_id=None):
+        super().__init__(conn_id=conn_id)
+        self._conn = None
+
+    def get_conn(self):
+        if self._conn is None:
+            if self._conn_id is None:
+                self._conn = s3fs.S3FileSystem()
+            else:
+                # TODO: Use same logic as existing S3/AWS hooks.
+                config = self.get_connection(self._conn_id)
+
+                extra_kwargs = {}
+                if 'encryption' in config.extra_dejson:
+                    extra_kwargs['ServerSideEncryption'] = \
+                        config.extra_dejson['encryption']
+
+                self._conn = s3fs.S3FileSystem(
+                    key=config.login,
+                    secret=config.password,
+                    s3_additional_kwargs=extra_kwargs)
+
+        return self._conn
+
+    def disconnect(self):
+        self._conn = None
+
+    def open(self, file_path, mode='rb'):
+        return self.get_conn().open(file_path, mode=mode)
+
+    def exists(self, file_path):
+        return self.get_conn().exists(file_path)
+
+    def isdir(self, path):
+        path = _remove_s3_prefix(path)
+
+        if self.sep not in path:
+            # Path is bucket name.
+            return True
+
+        parent_dir = self.dirname(path)
+
+        for child in self.get_conn().ls(parent_dir, detail=True):
+            if child['Key'] == path and \
+                    child['StorageClass'] == 'DIRECTORY':
+                return True
+
+        return False
+
+    def makedir(self, dir_path, mode=0o755, exist_ok=True):
+        self.makedirs(dir_path, mode=mode, exist_ok=exist_ok)
```

Comment:
S3 doesn't have the concept of directories -- you can just write to paths at will.

Reply:
That's why makedirs is essentially a no-op. It does check for existence of the 'directory', to keep the idea of a 'file system' in line with the other hooks.